### PR TITLE
[feat] Zeroize local variables on stack

### LINKED ISSUE #324: All Temp Data Must be Zeroized on Completion of Operation
Requirement: All Temp Data Must be Zeroized on Completion of Operation

Details:  Looking into rust, if we can prove that all data is zeroized upon garbage collection, then this will meet the requirement.  We will have to prove that this isn't disabled or able to work around.
Otherwise, before we leave scope for any intermediate data related to crypto operations, must be zeroized.

### PR DESCRIPTION
**Pull request by mhatrevi**
These changes are in response to the following FIPs requirement (https://github.com/chipsalliance/caliptra-sw/issues/324)

### CODE REVIEW

**Comment by JohnTraverAmd** [comment]
https://github.com/chipsalliance/caliptra-sw/pull/373

File: drivers/src/ecc384.rs [hunk 1]
```diff
@@ -109,6 +109,11 @@ impl Ecc384PubKey {
     pub fn to_der(&self) -> [u8; 97] {
         array_concat3([0x04], (&self.x).into(), (&self.y).into())
     }
+
+    pub fn zeroize(&mut self) {
```
Comment (korran) [inline drivers/src/ecc384.rs:113]: Have you looked at the machine code to see if this actually happens? I would expect the optimizer to remove this.

File: drivers/src/lms.rs [hunk 2]
```diff
@@ -366,6 +366,7 @@ impl Lms {
         }
         hasher.finalize()?;
         let result = HashValue::<N>::from(digest);
+        digest.0.fill(0);
```
Comment (korran) [inline drivers/src/lms.rs:369]: Why does it matter here? none of these values are secret....

File: rom/dev/src/flow/cold_reset/fmc_alias.rs [hunk 3]
```diff
@@ -70,14 +72,15 @@ impl FmcAliasLayer {
...
-        let output = DiceOutput {
+        let mut output = DiceOutput {
             subj_key_pair: key_pair,
             subj_sn,
             subj_key_id,
         };
 
         // Generate Local Device ID Certificate
         Self::generate_cert_sig(env, input, &output, fw_proc_info)?;
+        output.zeroize();
```
Comment (korran) [inline rom/dev/src/flow/cold_reset/fmc_alias.rs:83]: Are these really "temporaries"? Considering that they're being stored in the certificate, which isn't being zeroized, this seems silly to me.

Comment (mhatrevi) [reply rom/dev/src/flow/cold_reset/fmc_alias.rs:83]: Discussed in the Firmware standup; this is as per FIPS requirement (https://github.com/chipsalliance/caliptra-sw/issues/324)

File: drivers/src/lms.rs [hunk 4]
```diff
@@ -453,11 +454,14 @@ impl Lms {
             temp = HashValue::<N>::from(digest);
             node_num /= 2;
             i += 1;
+            digest.0.fill(0);
         }
-        let candidate_key = temp;
+        let mut candidate_key = temp;
```
Comment (korran) [inline drivers/src/lms.rs:459]: I think you probably want to clear temp here too, no?

I suggest:

```
let candidate_key = &temp;
if *candidate_key != HashValue::from(lms_public_key.digest) {
    return Ok(false);
}
digest.0.fill(0);
temp.fill(0);
```

File: rom/dev/src/flow/cold_reset/fmc_alias.rs [hunk 5]
```diff
@@ -95,9 +98,11 @@ impl FmcAliasLayer {
     fn derive_cdi(env: &mut RomEnv, measurements: Array4x12, cdi: KeyId) -> CaliptraResult<()> {
```
Comment (korran) [inline rom/dev/src/flow/cold_reset/fmc_alias.rs:98]: Should measurements be a reference and zeroized by the caller? (measurements cleared below is a separate shadowed variable)

File: drivers/src/lms.rs (hunk 4 above)
Comment (mhatrevi) [reply drivers/src/lms.rs:459]: Updated.

Comment (mhatrevi) [reply drivers/src/lms.rs]: Updated.

**Review by korran — APPROVED** [review]

**Pull request merged**